crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+pub struct MultiEd25519PublicKey(pub(crate) Vec<Ed25519PublicKey>, pub(crate) u8);
+
+/// Vector of the multi-key signatures along with their index required to map signatures with
+/// their corresponding public keys.
+#[derive(DeserializeKey, Clone, SerializeKey)]
+pub struct MultiEd25519Signature(pub(crate) Vec<Ed25519Signature>, pub(crate) BitVec);
+
+impl MultiEd25519PrivateKey {
+    /// Construct a new MultiEd25519PrivateKey.
+    pub fn new(
+        private_keys: Vec<Ed25519PrivateKey>,
+        threshold: u8,
+    ) -> std::result::Result<Self, CryptoMaterialError> {
+        let num_of_keys = private_keys.len();
+        if threshold != 0 && num_of_keys >= threshold as usize && num_of_keys <= MAX_NUM_OF_KEYS {
+            Ok(MultiEd25519PrivateKey(private_keys, threshold))
+        } else {
+            Err(CryptoMaterialError::WrongLengthError)
+        }
+    }
+
+    /// Serialize a MultiEd25519PrivateKey.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        to_bytes(&self.0, self.1)
+    }
+}
+
+impl MultiEd25519PublicKey {
+    /// Construct a new MultiEd25519PublicKey.
+    /// --- Rules ---
+    /// a) threshold cannot be zero.
+    /// b) public_keys.len() should be equal to or larger than threshold.
+    /// c) support up to MAX_NUM_OF_KEYS public keys.
+    pub fn new(
+        public_keys: Vec<Ed25519PublicKey>,
+        threshold: u8,
+    ) -> std::result::Result<Self, CryptoMaterialError> {
+        let num_of_keys = public_keys.len();
+        if threshold != 0 && num_of_keys >= threshold as usize && num_of_keys <= MAX_NUM_OF_KEYS {
+            Ok(MultiEd25519PublicKey(public_keys, threshold))
+        } else {
+            Err(CryptoMaterialError::WrongLengthError)
+        }
+    }
+
+    /// Serialize a MultiEd25519PublicKey.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        to_bytes(&self.0, self.1)
+    }
+}
+
+///////////////////////
+// PrivateKey Traits //
+///////////////////////
+
+impl PrivateKey for MultiEd25519PrivateKey {
+    type PublicKeyMaterial = MultiEd25519PublicKey;
+}
+
+impl SigningKey for MultiEd25519PrivateKey {
+    type VerifyingKeyMaterial = MultiEd25519PublicKey;
+    type SignatureMaterial = MultiEd25519Signature;
+
+    // Sign a message with the minimum amount of keys to meet threshold (starting from left-most keys).
+    fn sign_message(&self, message: &HashValue) -> MultiEd25519Signature {
+        let mut ed25519_signatures: Vec<Ed25519Signature> = Vec::with_capacity(self.1 as usize);
+        let mut bitvec = BitVec::from_elem(MAX_NUM_OF_KEYS, false);
+
+        for (i, item) in self.0.iter().enumerate() {
+            if i < self.1 as usize {
+                ed25519_signatures.push(item.sign_message(message));
+                bitvec.set(i, true);
+            } else {
+                break;
+            }
+        }
+        MultiEd25519Signature(ed25519_signatures, bitvec)
+    }
+}
+
+// Generating a random 2 out-of 3 key for testing.
```

Comment:
I'm not sure we should be generating group private keys, it seems an anti-pattern...

Reply:
It's only for testing purposes as mentioned in the comment.